pkg/validator/service_validator.go

```diff
@@ -72,6 +86,14 @@ func (sv *ServiceValidator) Handle(ctx context.Context, req admission.Request) a
 		return admission.Errored(http.StatusBadRequest, err)
 	}
 
+	if len(svc.Spec.ExternalIPs) > 0 && !sv.isValidUser(req.UserInfo) {
```

Comment:
Can you add unit test for Handle() to make sure its invoking isValidUser()? 

Also I would like to get @tallclair thought as well on whether this is the right model for restricting access to specific fields in API objects. Otherwise changes looks good to me.


> Users of this controller won't need to assign external IPs directly, instead users just create the controller's CRD. In this use case, only the service account for the controller is required to have the ability to assign external IPs.

Just to understand your use case better, can the controller create the service spec instead of allowing user to create one without external IP field? Then you can just use K8s RBAC.

Reply:
> Can you add unit test for Handle() to make sure its invoking isValidUser()?

Sure.

> Just to understand your use case better, can the controller create the service spec instead of allowing user to create one without external IP field? 

My use case is:
1. A user creates a service without external IP field,
2. A user defines a CRD for the controller,
3. The controller creates another service with external IP for the same endpoints(or selectors)
   (Or we might take an approach to directly modify the service that the user created to add external IP field).

So, the answer will be no.

> Then you can just use K8s RBAC and disallow users from creating the service spec.

Interesting. If the use case is to delegate all service creation to the controller and users only create services via the CRD for controller, it would be possible to restrict by using k8s RBAC. But my use case is to delegate service creation only for a specific purpose, so users will still need to create services for other purposes.